### PR TITLE
Update status based on admin approval

### PR DESCRIPTION
**Pull request by megcrenshaw**
This is the final PR for updating application statuses.

In https://github.com/code-dot-org/code-dot-org/pull/48949, the correct status is set when a teacher submits their application based on the RP's preferences.

In this PR, we update the status when (a) the admin approval is complete, and (b) the `principal_approval_not_required` field changes.

For reasons I don't understand, it looks like this PR has 22 commits (maybe it has to do with the squash merge?). But the Files Changed is correct, and the commits start at hash 00a56dfdc7a7f7c9caadf50912465b2361881043. Let me know if this is confusing.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/pd/application/principal_approval_application.rb [hunk 1]
```diff
@@ -42,12 +42,17 @@ class PrincipalApprovalApplication < ApplicationBase
...
+    after_create :update_teacher_app_status
 
     # @return a valid year (see Pd::SharedApplicationConstants::APPLICATION_YEARS)
     def year
       application_year
     end
 
+    def update_teacher_app_status
+      teacher_application.update!(status: 'unreviewed') if teacher_application.status == 'awaiting_admin_approval'
+    end
```
Comment (megcrenshaw) [inline dashboard/app/models/pd/application/principal_approval_application.rb:54]: When an application is in `awaiting_admin_approval`, it automatically becomes `unreviewed` as soon as the admin approval is received

File: dashboard/app/models/pd/application/teacher_application.rb [hunk 2]
```diff
@@ -135,6 +136,23 @@ def set_course_from_program
...
 
+    def set_status_from_admin_approval
+      # Do not modify status is application is not incomplete.
+      return if status == 'incomplete'
+
+      # Do not modify status if admin approval status has not changed.
+      # Since principal_approval_not_required is a serialized attribute, we cannot use the Dirty
+      # API easily –– instead, use the Dirty API to look at the properties attribute.
+      return if properties_change.include?(nil)
+      return unless properties_change.map(&:keys)&.flatten&.include?('principal_approval_not_required')
```
Comment (megcrenshaw) [inline dashboard/app/models/pd/application/teacher_application.rb:147]: This is very janky, but the only way I found to reliably see if the `principal_approval_not_required` field has changed. If you know of other approaches, I'm all ears!

File: dashboard/app/models/pd/application/principal_approval_application.rb [hunk 3]
```diff
@@ -42,12 +42,17 @@ class PrincipalApprovalApplication < ApplicationBase
...
                primary_key: :application_guid, foreign_key: :application_guid
+    after_create :update_teacher_app_status
 
     # @return a valid year (see Pd::SharedApplicationConstants::APPLICATION_YEARS)
     def year
       application_year
     end
 
+    def update_teacher_app_status
+      teacher_application.update!(status: 'unreviewed') if teacher_application.status == 'awaiting_admin_approval'
```
Comment (bethanyaconnor) [inline dashboard/app/models/pd/application/principal_approval_application.rb:53]: Are these statuses defined as constants anywhere? If not, that might be a nice clean up after launch.

File: dashboard/app/models/pd/application/teacher_application.rb (hunk 2 above)
Comment (bethanyaconnor) [reply dashboard/app/models/pd/application/teacher_application.rb:147]: This makes sense to me. Big downside of using the serialized attributes in this model :(

**Review by bethanyaconnor — APPROVED** [review]

**Comment by bethanyaconnor** [comment]
> For reasons I don't understand, it looks like this PR has 22 commits (maybe it has to do with the squash merge?). 

If you created this branch from another branch then squash merged the first branch, I think that would explain what you're seeing. I think these extra commits will end up being in the squash merge commit when you merge this PR but otherwise it shouldn't cause a problem (I think)

File: dashboard/app/models/pd/application/principal_approval_application.rb (hunk 3 above)
Comment (megcrenshaw) [reply dashboard/app/models/pd/application/principal_approval_application.rb:53]: Good call! I'll add it to the refactor list. I'm thinking I'll get good test coverage and then delve into refactors

**Comment by megcrenshaw** [comment]
> > For reasons I don't understand, it looks like this PR has 22 commits (maybe it has to do with the squash merge?).
> 
> If you created this branch from another branch then squash merged the first branch, I think that would explain what you're seeing. I think these extra commits will end up being in the squash merge commit when you merge this PR but otherwise it shouldn't cause a problem (I think)

Got it, thanks! That's what I figured.

File: dashboard/app/models/pd/application/teacher_application.rb (hunk 2 above)
Comment (megcrenshaw) [reply dashboard/app/models/pd/application/teacher_application.rb:147]: Amen to that

**Pull request merged**